shared/js/newtab/newtab.js

```diff
@@ -0,0 +1,130 @@
+import constants from '../../data/constants'
+import * as schema from './schema'
+const { events, allowedOrigin, clientPortName } = constants.trackerStats
+
+/**
+ * A flag to ensure we're not doing work until we've received a trusted event
+ * It's initially set to false, meaning we're inactive.
+ *
+ * Once a single `postMessage` is received from the allow-listed origin we
+ * will flip this switch and allow communications with the extension to continue
+ *
+ * @type {boolean}
+ */
+let receivedMessageFromAllowedOrigin = false
+
+/**
+ * Establish the connection with the extension, but not initially.
+ *
+ * Because this JS file is loaded into an iframe with access to `window.chrome.runtime`
+ * we use it to forward data into the parent page.
+ */
+function connect () {
+    const port = chrome.runtime.connect({ name: clientPortName })
+
+    port.onMessage.addListener((msg) => {
+        if (msg.messageType in events.outgoing) {
+            sendToNewTabPage(msg)
+        }
+    })
+
+    port.onDisconnect.addListener((msg) => {
+        console.log('🚨 PORT onDisconnect', msg)
+    })
```

Comment:
In MV3, if the service worker is shutdown, this port will disconnect, and the stats will stop updating for that new tab. Should we try to reconnect in that case?

Reply:
We'd have to reload the iframe either from within, or get the SERP side to detect the error and reload it.

short version is that you cannot re-connect once the port connection was lost (I tried lots of ways)